### PR TITLE
OCPBUGS-7714: Refactor & improve SyncWorker lock code

### PR DESCRIPTION
**Pull request by petr-muller**
Code refactors and cleanups that were previously part of https://github.com/openshift/cluster-version-operator/pull/1005, extracted to a separate PR because the actual OCPBUGS-7714 is possibly sensitive and we may want to review, discuss and investigate it (in the future) without the interference of the refactors.

Recommended to review per-commit.

- sync_worker.go: Refactor `calculateNext` and pull up lock
- sync_worker.go: add comments about lock assumptions
- sync_worker.go: lock access to w.status
- sync_worker.go: get rid of `StatusReporter` interface
- sync_worker.go: simplify `apply()` signature
- sync_worker.go: simplify `syncPayload` signature

### CODE REVIEW

**Comment by petr-muller** [comment]
/retest
:monocle_face:

**Review by wking — APPROVED** [review]
/lgtm

**Comment by petr-muller** [comment]
/label no-qe

Code refactors, no testable functionality. Regression testing is best performed by allowing this to merge and undergo full TRT and QE automation.

File: pkg/cvo/sync_worker.go
Comment (LalatenduMohanty) [inline pkg/cvo/sync_worker.go:668]: What is w.w ?

File: pkg/cvo/sync_worker.go
Comment (LalatenduMohanty) [inline pkg/cvo/sync_worker.go:678]: Copy paste error, s/ReportPayload/Report/

Comment (LalatenduMohanty) [inline pkg/cvo/sync_worker.go:678]: Copy paste error, s/ReportPayload/Report/

**Review by LalatenduMohanty — CHANGES_REQUESTED** [review]

File: pkg/cvo/sync_worker.go
Comment (petr-muller) [inline pkg/cvo/sync_worker.go:244]: And this is only partially true until #1005 merges, so I'll remove it from here and include it in the other PR

/lgtm cancel
/hold

File: pkg/cvo/sync_worker.go
Comment (petr-muller) [reply pkg/cvo/sync_worker.go:678]: will fix

Comment (petr-muller) [reply pkg/cvo/sync_worker.go]: fixed, also changed the godocs according to our Slack convo

File: pkg/cvo/sync_worker.go
Comment (petr-muller) [reply pkg/cvo/sync_worker.go]: done

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm

**Pull request merged**
